### PR TITLE
new feature level API for backends

### PR DESCRIPTION
**Pull request by pixelflinger**
backend can now return a "feature level", each level corresponds to a
"bundle" of features.
Level1: ES3.0 capabilities
Level2: ES3.1 capabilities + 31 textures + cubemap arrays

Currently metal always returns level 1, GL and Vulkan return level 2
if 31 textures or more are supported.

### CODE REVIEW

File: filament/backend/src/vulkan/VulkanDriver.h [hunk 1]
```diff
@@ -153,6 +153,7 @@ class VulkanDriver final : public DriverBase {
     VulkanSamplerGroup* mSamplerBindings[VulkanPipelineCache::SAMPLER_BINDING_COUNT] = {};
     VkDebugReportCallbackEXT mDebugCallback = VK_NULL_HANDLE;
     VkDebugUtilsMessengerEXT mDebugMessenger = VK_NULL_HANDLE;
+    FeatureLevel mFeatureLevel = FeatureLevel::FEATURE_LEVEL_1;
```
Comment (prideout) [inline filament/backend/src/vulkan/VulkanDriver.h:156]: NBD but I would not have bothered adding state here since we already persist the entire device caps and determining the answer to the query is a one liner.

Comment (pixelflinger) [reply filament/backend/src/vulkan/VulkanDriver.h:156]: ok, Engine is going to cache the result anyways. so yeah, I can change that.

Comment (pixelflinger) [reply filament/backend/src/vulkan/VulkanDriver.h]: done

File: filament/backend/include/backend/DriverEnums.h [hunk 2]
```diff
@@ -59,6 +59,15 @@ static_assert(MAX_VERTEX_BUFFER_COUNT <= MAX_VERTEX_ATTRIBUTE_COUNT,
 
 static constexpr size_t CONFIG_BINDING_COUNT = 12;  // This is guaranteed by OpenGL ES.
 
+/**
+ * Defines the backend's feature levels
```
Comment (bejado) [inline filament/backend/include/backend/DriverEnums.h:63]:
```suggestion
 * Defines the backend's feature levels.
```

File: filament/backend/src/opengl/OpenGLContext.cpp [hunk 3]
```diff
@@ -40,7 +40,7 @@ bool OpenGLContext::queryOpenGLVersion(GLint* major, GLint* minor) noexcept {
 OpenGLContext::OpenGLContext() noexcept {
     state.vao.p = &mDefaultVAO;
 
-    // There query work with all GL/GLES versions!
+    // These query work with all GL/GLES versions!
```
Comment (bejado) [inline filament/backend/src/opengl/OpenGLContext.cpp:43]:
```suggestion
    // These queries work with all GL/GLES versions!
```

**Review by bejado — APPROVED** [review]

**Comment by pixelflinger** [comment]
reviewers, I really recommend to look at the CLs one by one.

File: android/filament-android/src/main/cpp/Engine.cpp [hunk 4]
```diff
@@ -328,3 +328,24 @@ Java_com_google_android_filament_Engine_nIsAutomaticInstancingEnabled(JNIEnv*, j
     Engine* engine = (Engine*) nativeEngine;
     return (jboolean)engine->isAutomaticInstancingEnabled();
 }
+
+extern "C" JNIEXPORT jint JNICALL
+Java_com_google_android_filament_Engine_nGetSupportedFeatureLevel(JNIEnv *, jclass,
```
Comment (bejado) [inline android/filament-android/src/main/cpp/Engine.cpp:333]:
```suggestion
Java_com_google_android_filament_Engine_nGetSupportedFeatureLevel(JNIEnv*, jclass,
```

File: android/filament-android/src/main/cpp/Engine.cpp [hunk 5]
```diff
@@ -328,3 +328,24 @@ Java_com_google_android_filament_Engine_nIsAutomaticInstancingEnabled(JNIEnv*, j
...
+        jlong nativeEngine) {
+    Engine* engine = (Engine*) nativeEngine;
+    return (jint)engine->getSupportedFeatureLevel();
+}
+
+extern "C" JNIEXPORT jint JNICALL
+Java_com_google_android_filament_Engine_nSetActiveFeatureLevel(JNIEnv *, jclass,
```
Comment (bejado) [inline android/filament-android/src/main/cpp/Engine.cpp:340]:
```suggestion
Java_com_google_android_filament_Engine_nSetActiveFeatureLevel(JNIEnv*, jclass,
```

File: android/filament-android/src/main/cpp/Engine.cpp [hunk 6]
```diff
@@ -328,3 +328,24 @@ Java_com_google_android_filament_Engine_nIsAutomaticInstancingEnabled(JNIEnv*, j
...
+        jlong nativeEngine, jint ordinal) {
+    Engine* engine = (Engine*) nativeEngine;
+    return (jint)engine->setActiveFeatureLevel((Engine::FeatureLevel)ordinal);
+}
+
+extern "C" JNIEXPORT jint JNICALL
+Java_com_google_android_filament_Engine_nGetActiveFeatureLevel(JNIEnv *, jclass,
```
Comment (bejado) [inline android/filament-android/src/main/cpp/Engine.cpp:347]:
```suggestion
Java_com_google_android_filament_Engine_nGetActiveFeatureLevel(JNIEnv*, jclass,
```

File: docs/Materials.md.html [hunk 7]
```diff
@@ -932,6 +932,36 @@
...
+Type
+:    `number`
+
+Value
+:     An integer value, either 1 or 2. Defaults to 1.
+
+      Feature Level    |        Guaranteed features
+:----------------------|:---------------------------------
+1                      | 9 textures per material
+2                      | 12 textures per material, cubemap arrays, ESSL 3.10
```
Comment (bejado) [inline docs/Materials.md.html:946]: How did you come up with the 9 and 12 texture limits?

File: filament/src/components/RenderableManager.cpp [hunk 8]
```diff
@@ -253,6 +253,11 @@ RenderableManager::Builder::Result RenderableManager::Builder::build(Engine& eng
             continue;
         }
 
+        // we want a feature level violation to be a hard error (exception if enabled, or crash)
+        ASSERT_PRECONDITION(upcast(engine).hasFeatureLevel(material->getFeatureLevel()),
+                "Material \"%s\" has feature level %u which is not supported this Engine",
```
Comment (bejado) [inline filament/src/components/RenderableManager.cpp:258]: Maybe rephrase and add the current feature level of the engine:
```
Material xyz has feature level X which is not supported. This Engine instance supports feature level Y.
```
Same for `setMaterialInstanceAt`.

File: filament/src/details/Material.cpp [hunk 9]
```diff
@@ -45,32 +45,35 @@ namespace filament {
 using namespace backend;
 
 static MaterialParser* createParser(Backend backend, const void* data, size_t size) {
-    MaterialParser* materialParser = new MaterialParser(backend, data, size);
+    // unique_ptr do we don't leak MaterialParser on failures below
```
Comment (bejado) [inline filament/src/details/Material.cpp:48]:
```suggestion
    // unique_ptr so we don't leak MaterialParser on failures below
```

**Comment by bejado** [comment]
I think this is probably worth updating release notes?

File: docs/Materials.md.html (hunk 7 above)
Comment (pixelflinger) [reply docs/Materials.md.html:946]: 9 is what Impress is using now -- they use 9 and we're using 7. 
For feature level 2, I just picked 12 :-) I think we can revisit at some point.

Comment (bejado) [reply docs/Materials.md.html:946]: Okay, sounds good. Seems like if the backend supports 31 then we should be able to expose a lot more to materials.

Comment (pixelflinger) [reply docs/Materials.md.html:946]: Well, we need to keep some for us. And we're already technically "in debt". We could do 15 (users) + 16 (us). I think it's a limit we can raise later anyways, it wouldn't cause incompatibilities.

Comment (bejado) [reply docs/Materials.md.html:946]: 👍

File: android/filament-android/src/main/java/com/google/android/filament/Engine.java [hunk 10]
```diff
@@ -141,6 +142,18 @@ public enum Backend {
         NOOP,
     }
 
+    /**
+     * Defines the backend's feature levels.
+     */
+    public enum FeatureLevel {
+        /** Reserved, don't use */
+        FEATURE_LEVEL_0,
```
Comment (prideout) [inline android/filament-android/src/main/java/com/google/android/filament/Engine.java:150]: 😨

**Review by prideout — APPROVED** [review]
Feature Level 2 is documented in some places as "12 textures per material" and as "31 textures units" in other places. I get it, but do we need a comment somewhere (SibGenerator.h?) to explain the relationship, or to remind our future selves to update the docs when adding / removing samplers?
